### PR TITLE
Refactor learning rate scheduler

### PR DESCRIPTION
**Pull request by shihchengli**
## Description
Address #968 and #969. The major changes include:
1. Refactoring the learning rate scheduler with LambdaLR.
2. Changing the optimizer from Adam to AdamW, with the weight decay coefficient specified via CLI (defaulting to 0.01, the same as the AdamW default in [PyTorch](https://pytorch.org/docs/stable/generated/torch.optim.AdamW.html#torch.optim.AdamW)).
4. As discussed in #969, for infinite training, the cooldown steps will be 100 times the warmup steps.

## Questions
Do we want to rename the learning rate scheduler?

## Relevant issues
#968 and #969

## Checklist
- [x] linted with flake8?
- [ ] (if appropriate) unit tests added?

### CODE REVIEW

File: chemprop/models/model.py [hunk 1]
```diff
@@ -212,21 +217,44 @@ def predict_step(self, batch: TrainingBatch, batch_idx: int, dataloader_idx: int
...
+            },
+        ]
+
+        opt = optim.AdamW(optimizer_grouped_parameters, self.init_lr)
+        steps_per_epoch = self.trainer.num_training_batches
+        warmup_steps = self.warmup_epochs * steps_per_epoch
+        if self.trainer.max_epochs == -1:
+            warnings.warn(
+                "For infinite training, the number of cooldown epochs in learninig rate scheduler is set to 100 times the number of warmup epochs."
+            )
```
Comment (davidegraff) [inline chemprop/models/model.py:242]: this seems like a `logging.info` call

File: chemprop/models/model.py [hunk 2]
```diff
@@ -212,21 +217,44 @@ def predict_step(self, batch: TrainingBatch, batch_idx: int, dataloader_idx: int
...
+            cooldown_steps=cooldown_steps,
+            init_lr=self.init_lr,
+            max_lr=self.max_lr,
+            final_lr=self.final_lr,
         )
-        lr_sched_config = {
-            "scheduler": lr_sched,
-            "interval": "step" if isinstance(lr_sched, NoamLR) else "batch",
-        }
+        lr_sched = NoamLR(opt, **lr_kwargs)
```
Comment (davidegraff) [inline chemprop/models/model.py:255]: why group into a dictionary then unpack in the very next line?

File: chemprop/schedulers.py
Comment (davidegraff) [inline chemprop/schedulers.py:1]: Why are you subclassing the LR scheduler with what amounts to a parametrization of the superclass? This inheritance pattern seems pointless IMO

File: chemprop/models/model.py [hunk 3]
```diff
@@ -212,21 +217,44 @@ def predict_step(self, batch: TrainingBatch, batch_idx: int, dataloader_idx: int
...
+                ],
+                "weight_decay": self.weight_decay,
+            },
+            {
+                "params": [
+                    p for n, p in self.named_parameters() if any(nd in n for nd in no_decay)
+                ],
+                "weight_decay": 0.0,
+            },
+        ]
```
Comment (davidegraff) [inline chemprop/models/model.py:234]: Is there a particular reason some params don't decay? Also, do we have data showing that AdamW leads to measurable improvements? I don't like the thought of making the code more complicated for no practical benefit.

File: chemprop/models/model.py (hunk 2 above)
Comment (shihchengli) [reply chemprop/models/model.py]: I originally designed this for multiple learning rate schedulers, but I opted to use only the current learning rate scheduler, so I should just pass it directly.

File: chemprop/models/model.py (hunk 1 above)
Comment (shihchengli) [reply chemprop/models/model.py:242]: I use `warnings.warn` because this code is not in the CLI and, therefore, doesn't make use of the logger.

File: chemprop/schedulers.py
Comment (shihchengli) [reply chemprop/schedulers.py:1]: I have adjusted it into a function. Thank you for your comment.

File: chemprop/models/model.py (hunk 3 above)
Comment (shihchengli) [reply chemprop/models/model.py:234]: When using AdamW, it's generally recommended not to apply weight decay to bias and batch normalization parameters. Here's why:

1. Bias parameters: Bias terms are typically scalar values added to the output of a layer. Applying weight decay to bias terms can unnecessarily constrain the model's ability to fit the data. Excluding bias terms from weight decay allows the model to freely adjust these offsets without penalty.
2. Batch normalization parameters: BN layers have two learnable parameters: scale (weight) and shift (bias). Applying weight decay to these parameters can be counterproductive for several reasons:
      - BN parameters serve a different purpose than regular weights. They normalize and rescale activations rather than transform inputs
      - The scale parameter in BN is already constrained by the normalization process
      - Applying weight decay to BN parameters can lead to instability in training, causing periodic fluctuations in loss and accuracy (as discussed in this [post](https://towardsdatascience.com/weight-decay-and-its-peculiar-effects-66e0aee3e7b8))

Comment (shihchengli) [reply chemprop/models/model.py:234]: I haven't empirically tested AdamW in D-MPNN. I am happy to keep using Adam as it seems to be doing quite well in our case. I can also work on some benchmarks to compare them.

File: chemprop/schedulers.py [hunk 4]
```diff
@@ -1,11 +1,16 @@
...
 from torch.optim import Optimizer
-from torch.optim.lr_scheduler import LRScheduler
-
-
-class NoamLR(LRScheduler):
-    r"""A Noam learning rate scheduler schedules the learning rate with a piecewise linear followed
+from torch.optim.lr_scheduler import LambdaLR
+
+
+def get_NoamLR_sched(
```
Comment (shihchengli) [inline chemprop/schedulers.py:5]: I am not sure if this would be a better name
```suggestion
def get_linear_warmup_exponential_decay_sched(
```

File: chemprop/models/model.py (hunk 1 above)
Comment (davidegraff) [reply chemprop/models/model.py:242]: what do you mean? All code can make use of the logger. Look at lightning for an example

File: chemprop/models/model.py (hunk 3 above)
Comment (davidegraff) [reply chemprop/models/model.py:234]: 1. I'm less inclined to accept a reason when you give an answer from ChatGPT...
2. [This blog post](https://iclr-blogposts.github.io/2023/blog/2023/adamw/) seems to contradict (2).
3. It also mentions that AdamW generally performs slightly better or on-par with Adam for image tasks. I'd really have to see benchmarks on chemical tasks, as the optimizer does not seem to induce a level shift in performance

File: chemprop/schedulers.py (hunk 4 above)
Comment (davidegraff) [reply chemprop/schedulers.py:5]:
```suggestion
def build_NoamLike_LRSched(
```
I prefer `build` as the verb here. And we should update the docstring and name to reflect that this scheduler is "Noam-like" rather than actually a Noam scheduler.

File: chemprop/models/model.py (hunk 1 above)
Comment (shihchengli) [reply chemprop/models/model.py:242]: I made this statement because that is how we currently handle it in the codebase, also from [here](https://github.com/chemprop/chemprop/pull/953#issuecomment-2207250068). I am happy to always use the logger if it is the correct way to do it.

File: chemprop/models/model.py (hunk 3 above)
Comment (davidegraff) [reply chemprop/models/model.py:234]: In any case:

```suggestion
        no_decay_param_name = ["bias", "bn.weight"]
        decayed_params = dict(params=[], "weight_decay"=self.weight_decay)
        nondecayed_params = dict(params=[], "weight_decay"=0.0)
        for name, parameter in self.named_parameters():
            if any(no_decay_param_name in name for nd in no_decay):
                nondecayed_params["params"].append(parameter)
            else:
                decayed_params["params"].append(parameter)
        param_groups = [decayed_params, nondecayed_params]
```

Also, this seems like a fairly brittle approach using the `"bn.weight"` parameter name. But if it works it works...

File: chemprop/models/model.py (hunk 1 above)
Comment (davidegraff) [reply chemprop/models/model.py:242]: Good point!

File: chemprop/models/model.py (hunk 3 above)
Comment (shihchengli) [reply chemprop/models/model.py:234]: I have read some posts and feel many people have this question about whether putting weight decay on BN parameters would be helpful or harm the model. If you look at the blog post I showed above, it demonstrates that increasing epsilon to a larger value can resolve the issue of unstable training, so it might be task-dependent(?). For image recognition, I have also seen people don't use weight decay on bias and BN parameters. For example, this [paper](https://arxiv.org/abs/2102.06171) includes only the convolution and linear layer weights in weight decay. Similarly, in the [transformers](https://github.com/huggingface/transformers/tree/5f1fcc299cb00c1edce5eb1efb8bacdde2365690) package, they do not apply weight decay to the bias and LN parameters. This [discussion ](https://discuss.pytorch.org/t/weight-decay-in-the-optimizers-is-a-bad-idea-especially-with-batchnorm/16994) on the PyTorch Forum also has extensive discourse on why weight decay should not be applied to them. I also noticed a [paper](https://arxiv.org/abs/1706.05350) that studied why weight decay should not be used with batch normalization. Hope this is helpful.

Comment (davidegraff) [reply chemprop/models/model.py:234]: It is, and it's much better than relying on a ChatGPT answer with no references.

Comment (shihchengli) [reply chemprop/models/model.py:234]: I apologize if it seemed like I used ChatGPT to generate this without putting effort into thinking about the question. To clarify, I have read many posts and used Perplexity AI, which gives results backed by sources. I have read through these sources and believe they align with my thoughts, so I replied to you accordingly.

File: chemprop/schedulers.py (hunk 4 above)
Comment (shihchengli) [reply chemprop/schedulers.py]: Looks good! I have adopted it.

**Comment by shihchengli** [comment]
I would do some benchmarks on chemical tasks to see if AdamW would provide any benefit compared to use Adam.

**Review by davidegraff — APPROVED** [review]
If you can show that AdamW either increases convergence speed or final performance with some loss curves/final performance at a reasonable number of epochs, this looks good.

File: chemprop/models/model.py [hunk 5]
```diff
@@ -212,21 +217,33 @@ def predict_step(self, batch: TrainingBatch, batch_idx: int, dataloader_idx: int
...
+        nondecayed_params = dict(params=[], weight_decay=0.0)
+        for name, parameter in self.named_parameters():
+            if any(nd in name for nd in no_decay_param_name):
+                nondecayed_params["params"].append(parameter)
+            else:
+                decayed_params["params"].append(parameter)
+        param_groups = [decayed_params, nondecayed_params]
+
+        opt = optim.AdamW(param_groups, self.init_lr)
+        steps_per_epoch = self.trainer.num_training_batches
```
Comment (shihchengli) [inline chemprop/models/model.py:231]:
```suggestion
        steps_per_epoch = self.trainer.estimated_stepping_batches // self.trainer.max_epochs
```
`num_training_batches` doesn't work in `configure_optimizers`...
https://github.com/Lightning-AI/pytorch-lightning/issues/16060

**Comment by shihchengli** [comment]
Below are some benchmark results. Each model is trained with 50 epochs without batch normalization. There are no significant differences between using Adam and AdamW regarding the model performance and convergence speed. Based on these results, I would be happy to keep using the Adam optimizer.
QM9
image
ESOL:
image
FreeSolv:
image

File: chemprop/schedulers.py [hunk 6]
```diff
@@ -31,97 +36,30 @@ class NoamLR(LRScheduler):
...
-            self.optimizer.param_groups[i]["lr"] = self.lrs[i]
+    def lr_lambda(step: int):
+        if step < warmup_steps:
+            warmup_factor = (max_lr - init_lr) / warmup_steps
+            return step * warmup_factor / init_lr + 1
+        elif step > warmup_steps + cooldown_steps:
+            return final_lr / init_lr
+        else:
+            cooldown_factor = (final_lr / max_lr) ** (1 / cooldown_steps)
+            return (max_lr * (cooldown_factor ** (step - warmup_steps))) / init_lr
```
Comment (davidegraff) [inline chemprop/schedulers.py:63]:
```suggestion
        elif warmup_steps <= step < warmup_steps + cooldown_steps:
            cooldown_factor = (final_lr / max_lr) ** (1 / cooldown_steps)
            return (max_lr * (cooldown_factor ** (step - warmup_steps))) / init_lr
        else:
            return final_lr / init_lr
```
IMO it makes more sense stylistically to have the cases ordered like `x < a`, `a <= x < b`, and `b < x`

Comment (shihchengli) [reply chemprop/schedulers.py]: That is a good suggestion. Thanks!

File: chemprop/cli/train.py [hunk 7]
```diff
@@ -450,9 +450,16 @@ def process_train_args(args: Namespace) -> Namespace:
...
         )
+
     if args.output_dir is None:
         args.output_dir = Path(f"chemprop_training/{args.data_path.stem}/{NOW}")
 
+    if args.epochs != -1 and args.epochs <= args.warmup_epochs:
+        raise ArgumentError(
+            argument=None,
+            message=f"The number of epochs should be higher than the number of epochs during warmup. Got {args.epochs} epochs and {args.warmup_epochs} warmup epochs",
+        )
```
[the remark this answers is not in the document]
Comment (davidegraff) [reply chemprop/cli/train.py:461]: this should be handled at the level of the `Lightning.Trainer`'s via the `fast_dev_run` argument

[the remark this answers is not in the document]
Comment (KnathanM) [reply chemprop/cli/train.py:461]: So should we add another CLI argument to expose that to the user? I think it is simpler to let CLI users run for a single epoch. Though users can circumvent this argument error by `--warmup-epochs 1 --epochs 2`, so I guess I'm okay with leaving it all as is.

[the remark this answers is not in the document]
Comment (davidegraff) [reply chemprop/cli/train.py:461]: yeah IMO it would be more idiomatic to add a `-n/--dry-run` flag that runs only a few (2--4) batches under the hood _without_ writing any files (logs, checkpoints, tensorboard, output files, etc.)

[the remark this answers is not in the document]
Comment (davidegraff) [reply chemprop/cli/train.py:461]: I think if all of our outputs are stored under `output_dir`, then when the `--dry-run` flag is set, we can just override that to be `/dev/null`, silently writing everything to the ether.

[the remark this answers is not in the document]
Comment (shihchengli) [reply chemprop/cli/train.py:461]: I like the idea of have a `-n/--dry-run` flag for a quick test. @KnathanM we can discuss this in the dev meeting and decide whether to open an PR for this. As we already provided an error message here, users can circumvent the issue you mentioned for now.

File: chemprop/models/model.py (hunk 5 above)
Comment (KnathanM) [reply chemprop/models/model.py:221]: Does this need to be changed? I tested your branch and the training worked for me.

**Review by KnathanM — APPROVED** [review]

File: chemprop/models/model.py (hunk 5 above)
Comment (shihchengli) [reply chemprop/models/model.py:221]: I have solved this issue by the other method. Specifically, you can look at the L217-220.
